src/scratch.ts

```diff
@@ -20,21 +18,33 @@ export class ScratchDB extends DB {
    * Similar to `DB.get`, but first searches in-memory
    * scratch DB, if key not found, searches upstream DB.
    */
-  get(key: Buffer, cb: BufferCallback) {
-    const getDBs = this._upstream._leveldb
-      ? [this._leveldb, this._upstream._leveldb]
-      : [this._leveldb]
-    const dbGet = (db: any, cb2: Function) => {
-      db.get(key, ENCODING_OPTS, (err: Error, v: Buffer | null) => {
-        if (err || !v) {
-          cb2(null, null)
+  async get(key: Buffer): Promise<Buffer | null> {
+    let value = null
+    // First, search in-memory db
+    try {
+      value = await this._leveldb.get(key, ENCODING_OPTS)
+    } catch (error) {
+      if (error.notFound) {
+        // not found, returning null
+      } else {
+        throw error
+      }
+    }
+
+    // If not found, try searching upstream db
+    if (!value && this._upstream._leveldb) {
+      try {
+        value = await this._upstream._leveldb.get(key, ENCODING_OPTS)
+      } catch (error) {
+        if (error.notFound) {
+          // not found, returning null
         } else {
-          cb2(null, v)
+          throw error
         }
-      })
+      }
     }
 
-    asyncFirstSeries(getDBs, dbGet, cb)
+    return value
```

Comment:
Leave `scratch.ts` also open since not seeing through. Maybe will have a look with some fresh eyes in the following days myself, too tired right now, or otherwise open for a look from someone else as well (//cc @s1na @evertonfraga).

Reply:
Ok, finally got this whole construct decomposition. Phew. 😛 Extremely valuable to have this simplified on such a level, thanks Ryan for working yourself through stuff like this. This was really over-the-top complex and extremely hard to grasp before.